In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import io
import requests
url = "https://api.covidtracking.com/v1/states/daily.csv" # Access date and time : Feb 8, 2021, 10.30 A.M.
s=requests.get(url).content

In [3]:
df = pd.read_csv(io.StringIO(s.decode('utf-8')))

In [4]:
df.to_csv('../Dataset/dataset.csv', index=False, header=True) # Saving the dataset for reuse in future time

In [5]:
pd.set_option('display.max_columns', None) # To display all columns
df.head()

date state  positive  probableCases   negative  pending  \
0  20210207    AK   53279.0            NaN        NaN      NaN   
1  20210207    AL  472423.0       101367.0  1816273.0      NaN   
2  20210207    AR  306736.0        62862.0  2285451.0      NaN   
3  20210207    AS       0.0            NaN     2140.0      NaN   
4  20210207    AZ  780637.0        50509.0  2818265.0      NaN   

  totalTestResultsSource  totalTestResults  hospitalizedCurrently  \
0        totalTestsViral         1536911.0                   44.0   
1  totalTestsPeopleViral         2187329.0                 1513.0   
2        totalTestsViral         2529325.0                  781.0   
3        totalTestsViral            2140.0                    NaN   
4        totalTestsViral         6982148.0                 2910.0   

   hospitalizedCumulative  inIcuCurrently  inIcuCumulative  \
0                  1219.0             NaN              NaN   
1                 43005.0             NaN           2576.0   
2                 14066.0           270.0              NaN   
3                     NaN             NaN              NaN   
4                 54657.0           838.0              NaN   

   onVentilatorCurrently  onVentilatorCumulative  recovered dataQualityGrade  \
0                   11.0                     NaN        NaN                A   
1                    NaN                  1460.0   252880.0                A   
2                  126.0                  1458.0   285306.0               A+   
3                    NaN                     NaN        NaN              NaN   
4                  561.0                     NaN   107979.0               A+   

      lastUpdateEt          dateModified  checkTimeEt    death  hospitalized  \
0   2/5/2021 03:59  2021-02-05T03:59:00Z  02/04 22:59    279.0        1219.0   
1   2/7/2021 11:00  2021-02-07T11:00:00Z  02/07 06:00   8515.0       43005.0   
2   2/6/2021 00:00  2021-02-06T00:00:00Z  02/05 19:00   5076.0       14066.0   
3  12/1/2020 00:00  2020-12-01T00:00:00Z  11/30 19:00      0.0           NaN   
4   2/7/2021 00:00  2021-02-07T00:00:00Z  02/06 19:00  14048.0       54657.0   

            dateChecked  totalTestsViral  positiveTestsViral  \
0  2021-02-05T03:59:00Z        1536911.0             64404.0   
1  2021-02-07T11:00:00Z              NaN                 NaN   
2  2021-02-06T00:00:00Z        2529325.0                 NaN   
3  2020-12-01T00:00:00Z           2140.0                 NaN   
4  2021-02-07T00:00:00Z        6982148.0                 NaN   

   negativeTestsViral  positiveCasesViral  deathConfirmed  deathProbable  \
0           1470760.0                 NaN             NaN            NaN   
1                 NaN            371056.0          6747.0         1768.0   
2           2285451.0            243874.0          4054.0         1022.0   
3                 NaN                 0.0             NaN            NaN   
4                 NaN            730128.0         12480.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2187329.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3548393.0            423601.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109260.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19205 entries, 0 to 19204
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19205 non-null  int64  
 1   state                        19205 non-null  object 
 2   positive                     19016 non-null  float64
 3   probableCases                8489 non-null   float64
 4   negative                     15289 non-null  float64
 5   pending                      1999 non-null   float64
 6   totalTestResultsSource       19205 non-null  object 
 7   totalTestResults             19103 non-null  float64
 8   hospitalizedCurrently        15854 non-null  float64
 9   hospitalizedCumulative       11827 non-null  float64
 10  inIcuCurrently               10358 non-null  float64
 11  inIcuCumulative              3456 non-null   float64
 12  onVentilatorCurrently        8226 non-null   float64
 13  onVentilatorCumu

## Metadata information

1. __checkTimeEt__<br>
_Field type:string_<br>
Deprecated<br>

2. __commercialScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

3. __dataQualityGrade__<br>
Field type:string<br>
Data Quality Grade<br>
The COVID Tracking Project grade of the completeness of the data reporting by a state.<br>

4. __date__<br>
_Field type:integer_<br>
Date on which data was collected by The COVID Tracking Project.<br>

5. __dateChecked__<br>
Field type:string<br>
Deprecated<br>

6. __dateModified__<br>
_Field type:string_<br>
Deprecated, use lastUpdateEt instead<br>

7. __death__<br>
_Field type:integer_<br>
Deaths (confirmed and probable)<br>
Total fatalities with confirmed OR probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for fatalities, the metric includes only deaths with COVID-19 listed on the death certificate, unless deaths among cases is a more reliable metric in the state.<br>
_Returns : null. if no data is available_<br>

8. __deathConfirmed__<br>
_Field type:integer_<br>
Deaths (confirmed)<br>
Total fatalities with confirmed COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for confirmed fatalities, the metric includes only lab-confirmed deaths with COVID-19 listed on the death certificate, unless deaths among confirmed cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_

9. __deathIncrease__<br>
_Field type:integer_<br>
New deaths<br>
Daily increase in death, calculated from the previous day’s value.<br>
_Returns: null, if no data is available_<br>

10. __deathProbable__<br>
_Field type:integer_<br>
Deaths (probable)<br>
Total fatalities with probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for probable fatalities, the metric includes only probable fatalities with COVID-19 listed on the death certificate, unless deaths among probable cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_<br>

11. __fips__<br>
_Field type:string_<br>
FIPS code<br>
State FIPS code<br>

12. __grade__<br>
_Field type:string_<br>
Deprecated<br>

13. __hash__<br>
_Field type:string_<br>
Deprecated - A hash of the current record.<br>

14. __hospitalized__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

15. __hospitalizedCumulative__<br>
_Field type:integer_<br>
Cumulative hospitalized/Ever hospitalized<br>
Total number of individuals who have ever been hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_

16. __hospitalizedCurrently__<br>
_Field type:integer_<br>
Currently hospitalized/Now hospitalized<br>
Individuals who are currently hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

17. __hospitalizedIncrease__<br>
_Field type:integer_<br>
New total hospitalizations<br>
Daily increase in hospitalizedCumulative, calculated from the previous day’s value.<br>
_Returns : null, if no data is available_

18. __inIcuCumulative__<br>
_Field type:integer_<br>
Cumulative in ICU/Ever in ICU<br>
Total number of individuals who have ever been hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available<br>

19. __inIcuCurrently__<br>
_Field type:integer_<br>
Currently in ICU/Now in ICU<br>
Individuals who are currently hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

20. __lastUpdateEt__<br>
_Field type:string_<br>
Last Update (ET)<br>
Date and time in Eastern time the state or territory last updated the data.<br>

21. __negative__<br>
_Field type:integer_<br>
Negative PCR tests (people)<br>
Total number of unique people with a completed PCR test that returns negative. For states / territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. Due to complex reporting procedures, this number might be mixing units and therefore, at best, it should only be considered an estimate of the number of people with a completed PCR test that return negative.<br>
_Returns : null, if no data is available_

22. __negativeIncrease__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

23. __negativeRegularScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

24. __negativeScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

25. __negativeTestsAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (specimens)<br>
The total number of completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

26. __negativeTestsPeopleAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (people)<br>
The total number of unique people with completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

27. __negativeTestsViral__<br>
_Field type:integer_<br>
Negative PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return negative as reported by the state or territory. For states/territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

28. __onVentilatorCumulative__<br>
_Field type:integer_<br>
Cumulative on ventilator/Ever on ventilator<br>
Total number of individuals who have ever been hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

29. __onVentilatorCurrently__<br>
_Field type:integer_<br>
Currently on ventilator/Now on ventilator<br>
Individuals who are currently hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

30. __pending__<br>
_Field type:integer_<br>
Pending<br>
Total number of viral tests that have not been completed as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

31. __posNeg__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

32. __positive__<br>
_Field type:integer_<br>
Cases (confirmed plus probable)<br>
Total number of confirmed plus probable cases of COVID-19 reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions. Not all states and territories report probable cases. If a state is not reporting probable cases, this field will just represent confirmed cases.<br>
_Returns : null, if no data is available_<br>

33. __positiveCasesViral__<br>
_Field type:integer_<br>
Confirmed Cases<br>
Total number of unique people with a positive PCR or other approved nucleic acid amplification test (NAAT), as reported by the state or territory. This is equivalent to a confirmed case as per the CSTE case definitions of August 5th, 2020 and April 5th, 2020. If we discover a jurisdiction is labeling cases as confirmed using other evidence (e.g. positive antigen tests), we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

34. __positiveIncrease__<br>
_Field type:integer_<br>
New cases<br>
The daily increase in API field positive, which measures Cases (confirmed plus probable) calculated based on the previous day’s value.<br>
_Returns : null, if no data is available_<br>

35. __positiveScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

36. __positiveTestsAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (specimens)<br>
Total number of completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

37. __positiveTestsAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (specimens)<br>
Total number of completed antigen tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

38. __positiveTestsPeopleAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (people)<br>
The total number of unique people with completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

39. __positiveTestsPeopleAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (people)<br>
Total number of unique people with a completed antigen test that returned positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

40. __positiveTestsViral__<br>
_Field type:integer_<br>
Positive PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return positive as reported by the state or territory. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

41. __probableCases__<br>
_Field type:integer_<br>
Probable Cases<br>
Total number of probable cases of COVID-19 as reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. By this definition, a probable case is someone who tests positive via antigen without a positive PCR or other approved nucleic acid amplification test (NAAT), someone with clinical and epidemiological evidence of COVID-19 infection with no confirmatory laboratory testing performed for SARS-CoV-2, or someone with COVID-19 listed on their death certificate with no confirmatory laboratory testing performed for SARS-CoV-2. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions.<br>
_Returns : null, if no data is available_<br>

42. __recovered__<br>
_Field type:integer_<br>
Recovered<br>
Total number of people that are identified as recovered from COVID-19. States provide very disparate definitions on what constitutes a “recovered” COVID-19 case. Types of “recovered” cases include those who are discharged from hospitals, released from isolation after meeting CDC guidance on symptoms cessation, or those who have not been identified as fatalities after a number of days (30 or more) post disease onset. Specifics vary for each state or territory.<br>
_Returns : null, if no data is available_<br>

43. __score__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

44. __state__<br>
_Field type:string_<br>
State (or territory)<br>
Two-letter abbreviation for the state or territory.<br>

45. __total__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns :  null, if no data is available_

46. __totalTestEncountersViral__<br>
_Field type:integer_<br>
Total PCR tests (test encounters)<br>
Total number of people tested per day via PCR testing as reported by the state or territory. The count for this metric is incremented up by one for each day on which an individual person is tested, no matter how many specimens are collected from that person on that day. If an individual person is tested twice a day on three different days, this count will increment up by three. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

47. __totalTestResults__<br>
_Field type:integer_<br>
Total test results<br>
At the national level, this metric is a summary statistic which, because of the variation in test reporting methods, is at best an estimate of US viral (PCR) testing. Some states/territories report tests in units of test encounters, some report tests in units of specimens, and some report tests in units of unique people. Moreover, some jurisdictions include antigen tests in their total test counts without reporting a separate total of viral (PCR) tests. Therefore, this value is an aggregate calculation of heterogeneous figures. Please consult each state or territory’s individual data page to see whether that jurisdiction lumps antigen and PCR tests together and to see what units that jurisdiction uses for test reporting.<br>
In most states, the totalTestResults field is currently computed by adding positive and negative values because, historically, some states do not report totals, and to work around different reporting cadences for cases and tests. In Colorado, Delaware, the District of Columbia, Florida, Hawaii, Minnesota, Nevada, New York, North Dakota, Pennsylvania, Rhode Island, Virginia, Washington, and Wisconsin, where reliable testing encounters figures are available with a complete time series, we directly report those figures in this field. In Alaska, America Samoa, Arizona, Arkansas, California, Georgia, Indiana, Kentucky, Maryland, Massachusetts, Missouri, Nebraska, New Hampshire, Ohio, Oregon, Texas, Utah, Vermont, and Wyoming, where reliable specimens figures are available with a complete time series, we directly report those figures in this field. In Alabama, Idaho, and South Dakota, where reliable unique people figures are available with a complete time series, we directly report those figures in this field. We are in the process of switching all states over to use directly reported total figures, using a policy of preferring testing encounters, specimens, and people, in that order.<br>
_Returns : null, if no data is available_<br>


48. __totalTestResultsIncrease__<br>
_Field type:integer_<br>
New tests<br>
Daily increase in totalTestResults, calculated from the previous day’s value. This calculation includes all the caveats associated with Total tests/totalTestResults, and we recommend against using it at the state/territory level.<br>
_Returns : null, if no data is available_<br>

49. __totalTestResultsSource__<br>
_Field type:string_<br>
Indicates which field is being used for total test results. If it is posNeg, then it is calculated by adding all positive and negative values.<br>

50. __totalTestsAntibody__<br>
_Field type:integer_<br>
Total antibody tests (specimens)<br>
Total number of completed antibody tests as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

51. __totalTestsAntigen__<br>
_Field type:integer_<br>
Total antigen tests (specimens)<br>
Total number of completed antigen tests, as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

52. __totalTestsPeopleAntibody__<br>
_Field type:integer_<br>
Total antibody tests (people)<br>
The total number of unique people who have been tested at least once via antibody testing as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

53. __totalTestsPeopleAntigen__<br>
_Field type:integer_<br>
Total antigen tests (people)<br>
Total number of unique people who have been tested at least once via antigen testing, as reported by the state or territory.<br>
_Returns : null, if no data is available_

54. __totalTestsPeopleViral__<br>
_Field type:integer_<br>
Total PCR tests (people)<br>
Total number of unique people tested at least once via PCR testing, as reported by the state or territory. The count for this metric is incremented up only the first time an individual person is tested and their result is reported. Future tests of the same person will not be added to this count.<br>
In the case where the state only provides negative cases, this field is calculated as the summation of people who tested positive (“Positive Cases (People”) and the number of people who tested negative (“Negative (People or Cases)”). If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>


55. __totalTestsViral__<br>
_Field type:integer_<br>
Total PCR tests (specimens)<br>
Total number of PCR tests (or specimens tested) as reported by the state or territory. The count for this metric is incremented up each time a specimen is tested and the result is reported. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly. For states with ambiguous annotations, we have assigned their total tests to this category; these states and territories are identified in the new API field covidTrackingProjectPreferredTotalTestUnits as having “Unclear units.”<br>
_Returns : null, if no data is available_

In [7]:
df.corr() # Checking the correlation values between the different features

date  positive  probableCases  negative  \
date                         1.000000  0.403350       0.471940  0.356075   
positive                     0.403350  1.000000       0.901683  0.861454   
probableCases                0.471940  0.901683       1.000000  0.620788   
negative                     0.356075  0.861454       0.620788  1.000000   
pending                      0.050661  0.226807       0.708051  0.253257   
totalTestResults             0.343073  0.918836       0.760209  0.916254   
hospitalizedCurrently        0.182523  0.788657       0.801676  0.612512   
hospitalizedCumulative       0.236209  0.810323       0.717151  0.856888   
inIcuCurrently               0.109469  0.738103       0.791933  0.466253   
inIcuCumulative              0.376537  0.869864       0.751898  0.855113   
onVentilatorCurrently        0.130084  0.545098       0.473090  0.322949   
onVentilatorCumulative       0.463882  0.809749       0.893915  0.913105   
recovered                    0.363205  0.868807       0.867334  0.691471   
death                        0.270351  0.845571       0.707644  0.824111   
hospitalized                 0.236209  0.810323       0.717151  0.856888   
totalTestsViral              0.333125  0.938391       0.765405  0.925259   
positiveTestsViral           0.382847  0.985520       0.905840  0.867252   
negativeTestsViral           0.360361  0.959161       0.826980  0.952165   
positiveCasesViral           0.365034  0.996971       0.873507  0.848898   
deathConfirmed               0.320239  0.773162       0.434814  0.785706   
deathProbable                0.382624  0.587166       0.502899  0.539372   
totalTestEncountersViral     0.376650  0.906785       0.860444  0.983941   
totalTestsPeopleViral        0.317357  0.954197       0.771963  0.998758   
totalTestsAntibody           0.228938  0.902632       0.733022  0.865302   
positiveTestsAntibody        0.329251  0.873263       0.768252  0.743162   
negativeTestsAntibody        0.161242  0.913019       0.699025  0.920001   
totalTestsPeopleAntibody     0.154359  0.910430       0.621145  0.938031   
positiveTestsPeopleAntibody  0.285720  0.952671       0.716976  0.944991   
negativeTestsPeopleAntibody  0.228077  0.932197       0.763865  0.971803   
totalTestsPeopleAntigen      0.711857  0.841856       0.576764  0.820655   
positiveTestsPeopleAntigen   0.820228  0.916889       0.999773  0.779542   
totalTestsAntigen            0.534225  0.791388       0.925628  0.393414   
positiveTestsAntigen         0.599870  0.806190       0.900656  0.603656   
fips                         0.001356 -0.140387      -0.084427 -0.197303   
positiveIncrease             0.235618  0.802033       0.704384  0.682545   
negativeIncrease             0.094668  0.220117       0.108135  0.714452   
total                        0.344687  0.564814       0.458565  0.998705   
totalTestResultsIncrease     0.196783  0.784191       0.507602  0.737069   
posNeg                       0.344719  0.564790       0.458481  0.998714   
deathIncrease                0.207830  0.635918       0.642568  0.451441   
hospitalizedIncrease         0.040072  0.095773       0.107298  0.106330   
commercialScore                   NaN       NaN            NaN       NaN   
negativeRegularScore              NaN       NaN            NaN       NaN   
negativeScore                     NaN       NaN            NaN       NaN   
positiveScore                     NaN       NaN            NaN       NaN   
score                             NaN       NaN            NaN       NaN   
grade                             NaN       NaN            NaN       NaN   

                              pending  totalTestResults  \
date                         0.050661          0.343073   
positive                     0.226807          0.918836   
probableCases                0.708051          0.760209   
negative                     0.253257          0.916254   
pending                      1.000000          0.222137   
totalTestResults             0

In [9]:
# Checking the correlation between the death vs all the features only
df.corr()['death']

date                           0.270351
positive                       0.845571
probableCases                  0.707644
negative                       0.824111
pending                        0.207302
totalTestResults               0.852370
hospitalizedCurrently          0.657336
hospitalizedCumulative         0.941707
inIcuCurrently                 0.599082
inIcuCumulative                0.957400
onVentilatorCurrently          0.456948
onVentilatorCumulative         0.886804
recovered                      0.560364
death                          1.000000
hospitalized                   0.941707
totalTestsViral                0.866362
positiveTestsViral             0.935220
negativeTestsViral             0.952401
positiveCasesViral             0.873448
deathConfirmed                 0.997826
deathProbable                  0.803305
totalTestEncountersViral       0.854868
totalTestsPeopleViral          0.919917
totalTestsAntibody             0.947354
positiveTestsAntibody          0.874202


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19205 entries, 0 to 19204
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19205 non-null  int64  
 1   state                        19205 non-null  object 
 2   positive                     19016 non-null  float64
 3   probableCases                8489 non-null   float64
 4   negative                     15289 non-null  float64
 5   pending                      1999 non-null   float64
 6   totalTestResultsSource       19205 non-null  object 
 7   totalTestResults             19103 non-null  float64
 8   hospitalizedCurrently        15854 non-null  float64
 9   hospitalizedCumulative       11827 non-null  float64
 10  inIcuCurrently               10358 non-null  float64
 11  inIcuCumulative              3456 non-null   float64
 12  onVentilatorCurrently        8226 non-null   float64
 13  onVentilatorCumu

In [11]:
# Seems the total tests done = Positive + Negative
# However there are NAN values in negative and positive columns.
# positive = 19016
# negative = 15289
# total = 19205
# The data is lagging some positive cases, many negative cases
print(df['positive'].isna().sum())
print(df['negative'].isna().sum())
print(df['total'].isna().sum())

189
3916
0


In [12]:
# Some analysis on NAN values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]

# It shows that, if the negative value is nan then, the total equals positive

positive  negative    total
0        53279.0       NaN    53279
5      3335926.0       NaN  3335926
8        38035.0       NaN    38035
11      940991.0       NaN   940991
13       27163.0       NaN    27163
...          ...       ...      ...
19200        0.0       NaN        0
19201        0.0       NaN        0
19202        0.0       NaN        0
19203        0.0       NaN        0
19204        NaN       NaN        0

[3916 rows x 3 columns]

In [13]:
# Some analysis on NAN values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

# It shows that, if the negative value is nan then, the total equals positive

positive  negative  total
17587       NaN       NaN      0
17643       NaN       NaN      0
17699       NaN       NaN      0
17755       NaN       NaN      0
17811       NaN       NaN      0
...         ...       ...    ...
19188       NaN       NaN      0
19190       NaN       NaN      0
19192       NaN       NaN      0
19194       NaN       NaN      0
19204       NaN       NaN      0

[189 rows x 3 columns]

In [14]:
# Dropping all the columns with total = 0, which means the data is not available
df = df[df['total']!=0]

In [15]:
# Based on above filtering, checking out Nan values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

positive  negative  total
18577       NaN       NaN     14
18628       NaN       NaN     14
18679       NaN       NaN      9
18730       NaN       NaN      9
18781       NaN       NaN      6
18832       NaN       NaN      2
18883       NaN       NaN      1

In [16]:
# Based on above filtering, checking out Nan values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]

positive  negative    total
0        53279.0       NaN    53279
5      3335926.0       NaN  3335926
8        38035.0       NaN    38035
11      940991.0       NaN   940991
13       27163.0       NaN    27163
...          ...       ...      ...
19193        2.0       NaN        2
19195        2.0       NaN        2
19196        2.0       NaN        2
19197        1.0       NaN        1
19198        1.0       NaN        1

[3669 rows x 3 columns]

In [17]:
# Looking at the 3669 scenarios, all the positives = total when negative is nan, 
#thus replacing all the nan of negative with 0
df['negative'].fillna(0, inplace=True)

In [18]:
df[df['negative'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [19]:
# Placing all the values of total to positive, when the positive is nan
df['positive'] = df.apply(lambda row: row['total'] if np.isnan(row['positive']) else row['positive'], axis=1)

In [20]:
df[df['positive'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18953 entries, 0 to 19198
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         18953 non-null  int64  
 1   state                        18953 non-null  object 
 2   positive                     18953 non-null  float64
 3   probableCases                8489 non-null   float64
 4   negative                     18953 non-null  float64
 5   pending                      1997 non-null   float64
 6   totalTestResultsSource       18953 non-null  object 
 7   totalTestResults             18892 non-null  float64
 8   hospitalizedCurrently        15854 non-null  float64
 9   hospitalizedCumulative       11823 non-null  float64
 10  inIcuCurrently               10358 non-null  float64
 11  inIcuCumulative              3456 non-null   float64
 12  onVentilatorCurrently        8226 non-null   float64
 13  onVentilatorCumu

### Lets confirm, that the model, we want should be able to predict the death cases given the total tests done, and several other relevant features
### The relevant features that should be used for prediction are as follows:

1. Date, just to divide the test and train data 
2. state, its a categorical value, need to do one hot encoding on it, because based on the state info, the results could vary.
3. positive
4. negative
5. recovered
6. hospitalized
7. hospitalizedCumulative
8. hospitalizedCurrently
9. hospitalizedIncrease
10. inIcuCurrently
11. negativeIncrease
12. onVentilatorCumulative
13. onVentilatorCurrently
14. positiveCasesViral
15. positiveIncrease


_Most of the features are dropped based on the number of non-null counts, since, 
the total number of the values in such features are very less, thus appending zeros or other values in such features
will deviate the performance and the actual data relevancy, thus will try predicting on the basis of highly available 
feature columns and the relevant feature columns only._

------
<u>Target</u> is calculating the total **deaths** based on the above features.

In [22]:
# Selecting only the above mentioned features for further processing
selected_columns = ["date", "state", "positive", "negative", "recovered", "hospitalized", 
                    "hospitalizedCumulative", "hospitalizedCurrently", "hospitalizedIncrease", 
                    "inIcuCurrently", "negativeIncrease", "onVentilatorCurrently", 
                    "positiveCasesViral", "positiveIncrease", "total", "death"]

new_df = df[selected_columns]
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0  20210207    AK   53279.0        0.0        NaN        1219.0   
1  20210207    AL  472423.0  1816273.0   252880.0       43005.0   
2  20210207    AR  306736.0  2285451.0   285306.0       14066.0   
3  20210207    AS       0.0     2140.0        NaN           NaN   
4  20210207    AZ  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1219.0                   44.0                     0   
1                 43005.0                 1513.0                     0   
2                 14066.0                  781.0                    17   
3                     NaN                    NaN                     0   
4                 54657.0                 2910.0                   150   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                   11.0   
1             NaN              4462                    NaN   
2           270.0              8180                  126.0   
3             NaN                 0                    NaN   
4           838.0             16776                  561.0   

   positiveCasesViral  positiveIncrease    total    death  
0                 NaN                 0    53279    279.0  
1            371056.0              1112  2288696   8515.0  
2            243874.0               672  2592187   5076.0  
3                 0.0                 0     2140      0.0  
4            730128.0              1544  3598902  14048.0

In [23]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18953 entries, 0 to 19198
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   date                    18953 non-null  int64  
 1   state                   18953 non-null  object 
 2   positive                18953 non-null  float64
 3   negative                18953 non-null  float64
 4   recovered               13772 non-null  float64
 5   hospitalized            11823 non-null  float64
 6   hospitalizedCumulative  11823 non-null  float64
 7   hospitalizedCurrently   15854 non-null  float64
 8   hospitalizedIncrease    18953 non-null  int64  
 9   inIcuCurrently          10358 non-null  float64
 10  negativeIncrease        18953 non-null  int64  
 11  onVentilatorCurrently   8226 non-null   float64
 12  positiveCasesViral      13703 non-null  float64
 13  positiveIncrease        18953 non-null  int64  
 14  total                   18953 non-null

In [24]:
new_df.corr()['hospitalizedCumulative'] # hospitalized and hospitalizedCumulative are redundant features 

date                      0.236124
positive                  0.810348
negative                  0.391955
recovered                 0.337116
hospitalized              1.000000
hospitalizedCumulative    1.000000
hospitalizedCurrently     0.622000
hospitalizedIncrease      0.179922
inIcuCurrently            0.476767
negativeIncrease          0.155319
onVentilatorCurrently     0.551642
positiveCasesViral        0.919290
positiveIncrease          0.584562
total                     0.471509
death                     0.941700
Name: hospitalizedCumulative, dtype: float64

In [25]:
# dropping redundant features
new_df.drop('hospitalizedCumulative', inplace=True, axis=1)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [26]:
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0  20210207    AK   53279.0        0.0        NaN        1219.0   
1  20210207    AL  472423.0  1816273.0   252880.0       43005.0   
2  20210207    AR  306736.0  2285451.0   285306.0       14066.0   
3  20210207    AS       0.0     2140.0        NaN           NaN   
4  20210207    AZ  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   44.0                     0             NaN   
1                 1513.0                     0             NaN   
2                  781.0                    17           270.0   
3                    NaN                     0             NaN   
4                 2910.0                   150           838.0   

   negativeIncrease  onVentilatorCurrently  positiveCasesViral  \
0                 0                   11.0                 NaN   
1              4462                    NaN            371056.0   
2              8180                  126.0            243874.0   
3                 0                    NaN                 0.0   
4             16776                  561.0            730128.0   

   positiveIncrease    total    death  
0                 0    53279    279.0  
1              1112  2288696   8515.0  
2               672  2592187   5076.0  
3                 0     2140      0.0  
4              1544  3598902  14048.0

In [27]:
new_df['date'] =  pd.to_datetime(new_df['date'], format='%Y%m%d')

<ipython-input-27-214f810a4fa9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['date'] =  pd.to_datetime(new_df['date'], format='%Y%m%d')


In [28]:
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0 2021-02-07    AK   53279.0        0.0        NaN        1219.0   
1 2021-02-07    AL  472423.0  1816273.0   252880.0       43005.0   
2 2021-02-07    AR  306736.0  2285451.0   285306.0       14066.0   
3 2021-02-07    AS       0.0     2140.0        NaN           NaN   
4 2021-02-07    AZ  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   44.0                     0             NaN   
1                 1513.0                     0             NaN   
2                  781.0                    17           270.0   
3                    NaN                     0             NaN   
4                 2910.0                   150           838.0   

   negativeIncrease  onVentilatorCurrently  positiveCasesViral  \
0                 0                   11.0                 NaN   
1              4462                    NaN            371056.0   
2              8180                  126.0            243874.0   
3                 0                    NaN                 0.0   
4             16776                  561.0            730128.0   

   positiveIncrease    total    death  
0                 0    53279    279.0  
1              1112  2288696   8515.0  
2               672  2592187   5076.0  
3                 0     2140      0.0  
4              1544  3598902  14048.0

In [29]:
new_df["state"].unique()

array(['AK', 'AL', 'AR', 'AS', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL',
       'GA', 'GU', 'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA',
       'MD', 'ME', 'MI', 'MN', 'MO', 'MP', 'MS', 'MT', 'NC', 'ND', 'NE',
       'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'PR', 'RI',
       'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VI', 'VT', 'WA', 'WI', 'WV',
       'WY'], dtype=object)

### One hot encoding the states column

In [30]:
from sklearn.preprocessing import OneHotEncoder
# creating instance of one-hot-encoder
enc = OneHotEncoder(handle_unknown='ignore')
# passing state column (label encoded values of states)
enc_df = pd.DataFrame(enc.fit_transform(new_df[['state']]).toarray())
# merge with new_df on key values
new_df = new_df.join(enc_df)
new_df.head()
# enc_df

date state  positive   negative  recovered  hospitalized  \
0 2021-02-07    AK   53279.0        0.0        NaN        1219.0   
1 2021-02-07    AL  472423.0  1816273.0   252880.0       43005.0   
2 2021-02-07    AR  306736.0  2285451.0   285306.0       14066.0   
3 2021-02-07    AS       0.0     2140.0        NaN           NaN   
4 2021-02-07    AZ  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   44.0                     0             NaN   
1                 1513.0                     0             NaN   
2                  781.0                    17           270.0   
3                    NaN                     0             NaN   
4                 2910.0                   150           838.0   

   negativeIncrease  onVentilatorCurrently  positiveCasesViral  \
0                 0                   11.0                 NaN   
1              4462                    NaN            371056.0   
2              8180                  126.0            243874.0   
3                 0                    NaN                 0.0   
4             16776                  561.0            730128.0   

   positiveIncrease    total    death    0    1    2    3    4    5    6    7  \
0                 0    53279    279.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1              1112  2288696   8515.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
2               672  2592187   5076.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
3                 0     2140      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
4              1544  3598902  14048.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   

     8    9   10   11   12   13   14   15   16   17   18   19   20   21   22  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    23   24   25   26   27   28   29   30   31   32   33   34   35   36   37  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    38   39   40   41   42   43   44   45   46   47   48   49   50   51   52  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    53   54   55  
0  0.0  0.0  0.0  
1  0.0  0.0  0.0  
2  0.0  0.0  0.0  
3  0.0  0.0  0.0  
4  0.0  0.0  0.0

In [31]:
# Lets drop the state column now, since its already one hot encoded
new_df.drop('state', inplace=True, axis=1)

In [32]:
new_df.head()

date  positive   negative  recovered  hospitalized  \
0 2021-02-07   53279.0        0.0        NaN        1219.0   
1 2021-02-07  472423.0  1816273.0   252880.0       43005.0   
2 2021-02-07  306736.0  2285451.0   285306.0       14066.0   
3 2021-02-07       0.0     2140.0        NaN           NaN   
4 2021-02-07  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   44.0                     0             NaN   
1                 1513.0                     0             NaN   
2                  781.0                    17           270.0   
3                    NaN                     0             NaN   
4                 2910.0                   150           838.0   

   negativeIncrease  onVentilatorCurrently  positiveCasesViral  \
0                 0                   11.0                 NaN   
1              4462                    NaN            371056.0   
2              8180                  126.0            243874.0   
3                 0                    NaN                 0.0   
4             16776                  561.0            730128.0   

   positiveIncrease    total    death    0    1    2    3    4    5    6    7  \
0                 0    53279    279.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1              1112  2288696   8515.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
2               672  2592187   5076.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
3                 0     2140      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
4              1544  3598902  14048.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   

     8    9   10   11   12   13   14   15   16   17   18   19   20   21   22  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    23   24   25   26   27   28   29   30   31   32   33   34   35   36   37  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    38   39   40   41   42   43   44   45   46   47   48   49   50   51   52  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    53   54   55  
0  0.0  0.0  0.0  
1  0.0  0.0  0.0  
2  0.0  0.0  0.0  
3  0.0  0.0  0.0  
4  0.0  0.0  0.0

In [33]:
# Lets refill all the Nans with zeros
new_df.fillna(0, inplace=True)

In [34]:
new_df.head()

date  positive   negative  recovered  hospitalized  \
0 2021-02-07   53279.0        0.0        0.0        1219.0   
1 2021-02-07  472423.0  1816273.0   252880.0       43005.0   
2 2021-02-07  306736.0  2285451.0   285306.0       14066.0   
3 2021-02-07       0.0     2140.0        0.0           0.0   
4 2021-02-07  780637.0  2818265.0   107979.0       54657.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   44.0                     0             0.0   
1                 1513.0                     0             0.0   
2                  781.0                    17           270.0   
3                    0.0                     0             0.0   
4                 2910.0                   150           838.0   

   negativeIncrease  onVentilatorCurrently  positiveCasesViral  \
0                 0                   11.0                 0.0   
1              4462                    0.0            371056.0   
2              8180                  126.0            243874.0   
3                 0                    0.0                 0.0   
4             16776                  561.0            730128.0   

   positiveIncrease    total    death    0    1    2    3    4    5    6    7  \
0                 0    53279    279.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1              1112  2288696   8515.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
2               672  2592187   5076.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0   
3                 0     2140      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
4              1544  3598902  14048.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   

     8    9   10   11   12   13   14   15   16   17   18   19   20   21   22  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    23   24   25   26   27   28   29   30   31   32   33   34   35   36   37  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    38   39   40   41   42   43   44   45   46   47   48   49   50   51   52  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    53   54   55  
0  0.0  0.0  0.0  
1  0.0  0.0  0.0  
2  0.0  0.0  0.0  
3  0.0  0.0  0.0  
4  0.0  0.0  0.0

### Dividing data into train, and test sets

In [35]:
# Since the dataset is ordered in chronological manner, 
# lets put 70% of past data as training data and latest 30% as test data
mask = round(0.3*len(new_df))
print(mask)
test = new_df[:mask]
train = new_df[mask:]

5686


In [36]:
len(train)

13267

In [37]:
len(test)

5686

In [38]:
train.drop("date", inplace=True, axis=1)
test.drop('date', inplace=True, axis=1)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [39]:
# resetting the index 
train.reset_index(inplace=True, drop=True)

In [40]:
train_features = list(set(train.columns) - set(['death']))
# print(train_features)
X_train = train[train_features]
y_train = train["death"]
X_test = test[train_features]
y_test = test["death"]

### Simple Linear regression without any feature manipulations

In [41]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [42]:
simple_regressor = LinearRegression()
simple_regressor.fit(X_train, y_train)

LinearRegression()

In [43]:
predictions = simple_regressor.predict(X_train)

In [44]:
print('RMSE for Simple Linear Regression : ',np.sqrt(mean_squared_error(y_train, predictions)))

RMSE for Simple Linear Regression :  781.457450314224


#### Standardization of feature columns

In [49]:
from sklearn.preprocessing import StandardScaler

In [50]:
scaler = StandardScaler()

In [51]:
scaler.fit(X_train)

StandardScaler()

In [52]:
scaler.transform(X_train)

array([[-0.13370237, -0.13341003, -0.13457585, ...,  1.89592534,
        -0.33048884,  4.21786569],
       [-0.13370237, -0.13341003, -0.13457585, ..., -0.0849822 ,
        -0.33048884, -0.2120831 ],
       [-0.13370237, -0.13341003, -0.13457585, ...,  0.14745542,
        -0.33048884,  0.1298092 ],
       ...,
       [-0.13370237, -0.13341003, -0.13457585, ..., -0.45924014,
        -0.33048884, -0.53158853],
       [-0.13370237, -0.13341003, -0.13457585, ..., -0.45924925,
        -0.33048884, -0.53158853],
       [-0.13370237, -0.13341003, -0.13457585, ..., -0.45924925,
        -0.33048884, -0.53158853]])

In [53]:
standardized_regressor = LinearRegression()
standardized_regressor.fit(scaler.transform(X_train), y_train)

LinearRegression()

In [54]:
standardized_predictions = standardized_regressor.predict(scaler.transform(X_train))

In [55]:
print('RMSE for Standardized Linear Regression : ',np.sqrt(mean_squared_error(y_train, standardized_predictions)))

RMSE for Standardized Linear Regression :  781.4574503142238


Clearly, the above result of RMSE is very bad, saying that the model is underfitting

-------

#### Trying out polynomial features for fitting the data

In [56]:
from sklearn.preprocessing import PolynomialFeatures

In [59]:
poly_2 = PolynomialFeatures(degree = 2) 
X_poly_2 = poly_2.fit_transform(X_train)

poly_scaler_2 = StandardScaler()
poly_scaler_2.fit(X_poly_2)

poly_reg_2 = LinearRegression() 
poly_reg_2.fit(poly_scaler_2.transform(X_poly_2), y_train) 
pred_2 = poly_reg_2.predict(poly_scaler_2.transform(X_poly_2))

print('RMSE for polynomial features with degree 2 Regression :', np.sqrt(mean_squared_error(y_train, pred_2)))

RMSE for polynomial features with degree 2 Regression : 80.83447281126186


Degree 2 for the polynomial feature has reduced the RMSE value by almost 10 times, the model is obviously not underfitting. 

----

In [60]:
poly_3 = PolynomialFeatures(degree = 3) 
X_poly_3 = poly_3.fit_transform(X_train)

poly_scaler_3 = StandardScaler()
poly_scaler_3.fit(X_poly_3)

poly_reg_3 = LinearRegression() 
poly_reg_3.fit(poly_scaler_3.transform(X_poly_3), y_train) 
pred_3 = poly_reg_3.predict(poly_scaler_3.transform(X_poly_3))

print('RMSE for polynomial features with degree 3 Regression :', np.sqrt(mean_squared_error(y_train, pred_3)))

RMSE for polynomial features with degree 3 Regression : 17.07924163175371


### Result analysis

The linear regression with polynomial feature has decreased the RMSE error significantly.
The degree 3 polynomial feature has the lowest RMSE in training data, that is 17.079.

In [79]:
# Lets check if the model is overfitting.
print("RMSE for test data with polynomial regression of degree 3 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE for test data with polynomial regression of degree 3 :  7.659619838501448e+16


It can be clearly seen from the RMSE score that the polynomial regression model of degree 3 is overfitting alot

In [81]:
# Lets check if the model is overfitting in polynomial regression with degree 2.
test_poly_2 = PolynomialFeatures(degree = 2) 
X_test_2 = test_poly_2.fit_transform(X_test)
print("RMSE for test data with polynomial regression of degree 2 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE for test data with polynomial regression of degree 2 :  106669159937314.16


Polynomial regression model of degree 2 is also very overfitting

In [82]:
print("RMSE for test data with standardized regressor : ", np.sqrt(mean_squared_error(y_test, standardized_regressor.predict(scaler.transform(X_test)))))

RMSE for test data with standardized regressor :  3025.8240038989434


------

### Applying regularization

#### L2 regularization

#### L2 regularization on linear regression

In [161]:
from sklearn.linear_model import Ridge
ridge_standardized = Ridge(alpha=100.0)

In [162]:
ridge_standardized.fit(scaler.transform(X_train), y_train)

Ridge(alpha=100.0)

In [163]:
print('RMSE on training data for Standardized Ridge Regression : ',np.sqrt(mean_squared_error(y_train, ridge_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized Ridge Regression :  790.8509437430997


In [164]:
print("RMSE for test data with ridge regressor : ", np.sqrt(mean_squared_error(y_test, ridge_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with ridge regressor :  2819.4099616242806


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result.

#### L2 regularization on polynomial regression of degree 2

In [165]:
ridge_poly_2 = Ridge(alpha=100.0)

In [166]:
ridge_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

Ridge(alpha=100.0)

In [167]:
print('RMSE on training data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression :  156.11568778186938


In [168]:
print('RMSE on test data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on training data for Standardized Ridge Regression :  4272.891237210047


By adding L2 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 4272. The addition of L2 regularization on the model has scaled the test error from factor of 10^15 to 1K, which is a huge improvement interms of the model generalizability.

#### L2 regularization on polynomial regression of degree 3

In [165]:
ridge_poly_3 = Ridge(alpha=100.0)

In [169]:
ridge_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

Ridge()

In [170]:
print('RMSE on training data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression with degree 3 :  28.74896331312586


In [171]:
print('RMSE on test data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Ridge Regression with degree 3 :  17804.929700532004


By adding L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 17804. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 regularization

#### L1 regularization on linear regression

In [180]:
from sklearn.linear_model import Lasso
lasso_standardized = Lasso(alpha=100.0)

In [181]:
lasso_standardized.fit(scaler.transform(X_train), y_train)

Lasso(alpha=100.0)

In [182]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, lasso_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized Ridge Regression :  953.3144018862494


In [183]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, lasso_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with ridge regressor :  3070.182596553438


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1 or L2 regression are quiet comparable

#### L1 regularization on polynomial regression of degree 2

In [184]:
lasso_poly_2 = Lasso(alpha=100.0)

In [185]:
lasso_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [186]:
print('RMSE on training data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 2 :  607.7821862627763


In [187]:
print('RMSE on test data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 2 :  4599.883232362514


By adding L1 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 4599. The addition of L1 regularization on the model has scaled the test error from factor of 10^15 to 1K, which is a huge improvement interms of the model generalizability.

#### L1 regularization on polynomial regression of degree 3

In [188]:
lasso_poly_3 = Lasso(alpha=100.0)

In [189]:
lasso_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [190]:
print('RMSE on training data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 3 :  598.5867956501127


In [191]:
print('RMSE on test data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 3 :  5682.515308429499


By adding L1 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 5682. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3. The L1 regularization has shown better generalizibility than the L2 regularization for the polynomial regression of order 3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 and L2 regularization

#### L1/L2 regularization on linear regression

In [192]:
from sklearn.linear_model import ElasticNet

In [200]:
elastic_standardized = ElasticNet(alpha=1.0)

In [201]:
elastic_standardized.fit(scaler.transform(X_train), y_train)

ElasticNet()

In [202]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, elastic_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized lasso Regression :  1191.3545334885946


In [203]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, elastic_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with lasso regressor :  2419.857338129787


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1, L2, regularization are quiet comparable. However, the L1/L2 regularization with very less penalty has very good performance.

#### L1/L2 regularization on polynomial regression of degree 2

In [212]:
elastic_poly_2 = ElasticNet(alpha=1.0)

In [213]:
elastic_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

ElasticNet()

In [216]:
print('RMSE on training data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 2 :  498.7737480143985


In [217]:
print('RMSE on test data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 2 :  9373.648415816955


By adding L1/L2 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 9373. The addition of L1/L2 regularization on the model has scaled the test error from factor of 10^15 to 10K, which is a huge improvement interms of the model generalizability. However, the model performance is not par as compared to the L1 only and L2 only regularization.

#### L1/L2 regularization on polynomial regression of degree 3

In [218]:
elastic_poly_3 = ElasticNet(alpha=1.0)

In [219]:
elastic_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 2448670537.5238323, tolerance: 22177518.87666834
  model = cd_fast.enet_coordinate_descent(


ElasticNet()

In [220]:
print('RMSE on training data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 3 :  346.99643131125737


In [221]:
print('RMSE on test data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 3 :  18681.878235384993


By adding L1/L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 18681. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^4.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

#### Lastly, the addition of the regularization has improved the model generalazibility, plus improved the training time by alot. However, the model performance is not as expected, but impact of regularization can be clearly seen from the results above. The reason behind the bad model performance is because of the data. The NAN value filled with zeros is one of the reason why the model was not able to perform well. Since, introducing zeros introduces a noise in the data, which means the data behavior is completely changed due to it.